contracts/Scroll_SpokePool.sol

```diff
@@ -0,0 +1,123 @@
+     * @param _l2ScrollMessenger Scroll Messenger contract on L2.
+     * @param _initialDepositId Starting deposit ID. Set to 0 unless this is a re-deployment in order to mitigate
+     * @param _crossDomainAdmin Cross domain admin to set. Can be changed by admin.
+     * @param _hubPool Hub pool address to set. Can be changed by admin.
+     */
+    function initialize(
+        IL2GatewayRouter _l2GatewayRouter,
+        IScrollMessenger _l2ScrollMessenger,
+        uint32 _initialDepositId,
+        address _crossDomainAdmin,
+        address _hubPool
+    ) public initializer {
+        __SpokePool_init(_initialDepositId, _crossDomainAdmin, _hubPool);
+        l2GatewayRouter = _l2GatewayRouter;
+        l2ScrollMessenger = _l2ScrollMessenger;
+    }
+
+    /**
+     * @notice Change the L2 Gateway Router. Changed only by admin.
+     * @param _l2GatewayRouter New address of L2 gateway router.
+     */
+    function setL2GatewayRouter(IL2GatewayRouter _l2GatewayRouter) public onlyAdmin nonReentrant {
+        _setL2GatewayRouter(_l2GatewayRouter);
+    }
+
+    /**
+     * @notice Change L2 message service address. Callable only by admin.
+     * @param _l2ScrollMessenger New address of L2 messenger.
+     */
+    function setL2ScrollMessenger(IScrollMessenger _l2ScrollMessenger) public onlyAdmin nonReentrant {
+        _setL2MessageService(_l2ScrollMessenger);
+    }
+
+    /**************************************
+     *         INTERNAL FUNCTIONS         *
+     **************************************/
+
+    /**
+     * @notice Bridge tokens to the HubPool.
+     * @param amountToReturn Amount of tokens to bridge to the HubPool.
+     * @param l2TokenAddress Address of the token to bridge.
+     */
+    function _bridgeTokensToHubPool(uint256 amountToReturn, address l2TokenAddress) internal virtual override {
+        // The scroll bridge handles arbitrary ERC20 tokens and is mindful of
+        // the official WETH address on-chain. We don't need to do anything specific
+        // to differentiate between WETH and a separate ERC20.
+        // Note: This happens due to the L2GatewayRouter.getERC20Gateway() call
+        l2GatewayRouter.withdrawERC20(l2TokenAddress, hubPool, amountToReturn, 0);
+        emit ScrollTokensBridged(l2TokenAddress, hubPool, amountToReturn);
+    }
+
+    /**
+     * @notice Verifies that calling method is from the cross domain admin.
+     */
+    function _requireAdminSender() internal view override {
+        // The xdomainMessageSender is set within the Scroll messenger right
+        // before the call to this function (and reset afterwards). This represents
+        // the address that sent the message from L1 to L2. If the calling contract
+        // isn't the Scroll messenger, then the xdomainMessageSender will be the zero
+        // address and *NOT* cross domain admin.
+        address _xDomainSender = l2ScrollMessenger.xDomainMessageSender();
```

Comment:
Can xDomainMessageSender be set to whatever a malicious contract wants it to or is it always the contract that is the `target` of the L1 to L2 message?

Reply:
(I'm going to go through each call in the stack so this might be longform, but tl;dr I believe we should be fine as long as we trust the chain)

---

Per master on `scroll`'s official github ( [link](https://github.com/scroll-tech/scroll/blob/develop/contracts/src/L2/L2ScrollMessenger.sol#L142-L160) ), the `xDomainMessenger` is set to be the `_from` sender (sender of the L1 txn) here. The transaction stack is `l2TokenMessenger` calls `relayMessage` -> `_executeMessage`.


### _executeMessage
```
function _executeMessage(
        address _from,
        address _to,
        uint256 _value,
        bytes memory _message,
        bytes32 _xDomainCalldataHash
    ) internal {
        // @note check more `_to` address to avoid attack in the future when we add more gateways.
        require(_to != messageQueue, "Forbid to call message queue");
        _validateTargetAddress(_to);

        // @note This usually will never happen, just in case.
        require(_from != xDomainMessageSender, "Invalid message sender"); 
```
the last line of this snippet always set to be the DEFAULT_XDOMAIN_MESSAGE_SENDER outside of this specific case. This is a call to ensure that doesn't happen

```
        xDomainMessageSender = _from;
        // solhint-disable-next-line avoid-low-level-calls
        (bool success, ) = _to.call{value: _value}(_message);
        // reset value to refund gas.
        xDomainMessageSender = ScrollConstants.DEFAULT_XDOMAIN_MESSAGE_SENDER; <--- Immediately resets
        ...
    }
```
In the above ^ it's only set to be read within the context of `_to.call(...)` and then is immediately reset

### relayMessage

```
    function relayMessage(
        address _from,
        address _to,
        uint256 _value,
        uint256 _nonce,
        bytes memory _message
    ) external override whenNotPaused {
        // It is impossible to deploy a contract with the same address, reentrance is prevented in nature.
        require(AddressAliasHelper.undoL1ToL2Alias(_msgSender()) == counterpart, "Caller is not L1ScrollMessenger");

        bytes32 _xDomainCalldataHash = keccak256(_encodeXDomainCalldata(_from, _to, _value, _nonce, _message));

        require(!isL1MessageExecuted[_xDomainCalldataHash], "Message was already successfully executed");

        _executeMessage(_from, _to, _value, _message, _xDomainCalldataHash);
    }
```

From the above, what we're really looking for is that `_msgSender()` which is the `msg.sender` is the counterpart domain of the `L1ScrollMessenger`. Ergo if we pass that require, we're on the `L2ScrollMessenger`.

### L2ScrollMessenger

We have an implicit trust that the `L2ScrollMessenger` is going to set things right.